components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthnService.java

```diff
@@ -258,4 +295,245 @@ private boolean canAuthenticate(OAuthClientAuthenticator oAuthClientAuthenticato
+     *
+     * @param request                Http servlet request.
+     * @param contentParams          Map of request body params.
+     * @return Client authentication method of the request.
+     * @throws OAuthClientAuthnException
+     */
+    public String retrieveRequestAuthMethod(HttpServletRequest request, Map<String, List> contentParams) throws
+            OAuthClientAuthnException {
+
+        try {
+            if (isPrivateKeyJWTAuthentication(request)) {
+                log.debug("Validating request with JWT client authentication method");
+                return OAuthConstants.PRIVATE_KEY_JWT;
+            } else if (isMTLSAuthentication(request, contentParams)) {
+                log.debug("Validating request with MTLS client authentication method");
+                return OAuthConstants.TLS_CLIENT_AUTH;
+            }
+            return "INVALID_AUTH";
+        } catch (OAuthClientAuthnException e) {
+            throw new OAuthClientAuthnException("Error occurred when obtaining request authentication method",
+                    OAuth2ErrorCodes.INVALID_REQUEST, e);
+        }
+    }
+
+    /**
+     * Obtain the client authentication method registered for the application.
+     *
+     * @param clientId     Client ID of the application.
+     * @return Registered client authentication method for the application.
+     * @throws OAuthClientAuthnException
+     */
+    public String retrieveRegisteredAuthMethod(String clientId) throws OAuthClientAuthnException {
+
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            if (StringUtils.isEmpty(serviceProvider.getCertificateContent())) {
+                ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+                for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+                    if (OAuthConstants.TOKEN_ENDPOINT_AUTH_METHOD.equals(serviceProviderProperty.getName())) {
+                        return serviceProviderProperty.getValue();
+                    }
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException("Token signing algorithm not registered",
+                    OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+        return OAuthConstants.NOT_APPLICABLE;
+    }
+
+    /**
+     * Check whether the application should support FAPI.
+     *
+     * @param clientId       Client ID of the application.
+     * @return Whether the application should support FAPI.
+     * @throws OAuthClientAuthnException
+     */
+    private String isFapiApplication(String clientId) throws OAuthClientAuthnException {
+
+        String isFapiApp = "IsFAPIApp";
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            if (StringUtils.isEmpty(serviceProvider.getCertificateContent())) {
```

Comment:
Wondering why we want to evaluate if the certificate is empty for this logic.

Reply:
Fixed in 2604e0460788414d7a215802ac7a3ec386fb6fda